Input/Core/InputEventPool.cs

```diff
@@ -1,26 +1,60 @@
 using System;
 using System.Collections.Generic;
-using UnityEngine;
 
 namespace UnityEngine.InputNew
 {
 	class InputEventPool
 	{
-		#region Public Methods
+		readonly Dictionary<Type, List<InputEvent>> m_Pools = new Dictionary<Type, List<InputEvent>>();
 
+		#region Public Methods
 		public TEvent ReuseOrCreate<TEvent>()
 			where TEvent : InputEvent, new()
 		{
-			////TODO
+			var enumerator = m_Pools.GetEnumerator();
+			while (enumerator.MoveNext())
+			{
+				var kvp = enumerator.Current;
+				if (kvp.Key == typeof(TEvent))
+				{
+					var pool = kvp.Value;
+					if (pool.Count > 0)
+					{
+						var last = pool.Count - 1;
+						var inputEvent = pool[last];
+						pool.RemoveAt(last);
+						inputEvent.Reset();
+						return (TEvent)inputEvent;
+					}
+				}
+			}
+			enumerator.Dispose();
+
 			return new TEvent();
 		}
 
 		public void Return(InputEvent inputEvent)
 		{
-			////TODO
-			inputEvent.Reset();
-		}
+			var type = inputEvent.GetType();
 
+			List<InputEvent> pool = null;
+			var enumerator = m_Pools.GetEnumerator();
+			while (enumerator.MoveNext())
+			{
+				var kvp = enumerator.Current;
+				if (kvp.Key == type)
+					pool = kvp.Value;
+			}
+			enumerator.Dispose();
```

Comment:
The while(enumerator) thing actually only applies to <5.6, so we could revert it if we want.

Same goes for the foreach -> for. Would be nice to keep it in case people are using this in legacy versions

Reply:
This is the 2017 branch, so we can do what we want here